library/alloc/src/collections/btree/borrow.rs

```diff
@@ -0,0 +1,42 @@
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+
+/// A mutable borrow of a type T, that allows intermediately borrowing part of itself,
+/// until waking up the borrow of the entire T.
+pub struct DormantMutRef<'t, T> {
+    ptr: NonNull<T>,
+    _marker: PhantomData<&'t mut T>,
+}
+
+impl<'t, T> DormantMutRef<'t, T> {
+    /// Capture a unique borrow and immediately sublet it to something else,
+    /// that contains one or more unique references to a part of the T object.
+    /// `compute_sub` returns that something, starting from a pointer.
+    /// `compute_sub`, and all code using the returned something later, must not
+    /// dereference that pointer entirely but only one or more of its fields.
+    /// It's wise to store the returned`Self` and `Sub` in the same object,
+    /// so that they are consumed jointly.
+    pub fn new<Sub, ComputeSub>(t: &'t mut T, compute_sub: ComputeSub) -> (Self, Sub)
+    where
+        ComputeSub: FnOnce(NonNull<T>) -> Sub,
```

Comment:
Is there ever a case where I cannot just make `compute_sub` the identity function, and to the subsequent computations after calling `new` here? I am wondering why there is a closure.

Reply:
Sure you can, but then why are you using this code at all? It's meant to help you do this kind of thing safely, keeping the raw pointer(s) as much encapsulated as possible. The closure is there to get back to a "checked" reference (to a part) as quickly as possible.

[Originally](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=43dce4f7b9e384a18fe8759a2e7761f9), the "Sub" was part of the "Self" and `awaken` (or `terminate`) was not unsafe but it guaranteed itself (as far as I understand borrowing) that "Sub" was consumed. I'm trying to remember why I had to let go of that.

Reply:
Oh so the point is that the `NonNull` does not have the closure. Makes sense, but please document this.

Reply:
Now I am confused... how can a `NonNull` "have" a closure? It's one brain cell up from a dumb pointer??

Reply:
I meant "leave", sorry.

Reply:
Having "Sub" as a part of "Self" has two disadvantages:
- The "sub" or "part" field has a meaningless name. Although in practice the same field is now mostly called "handle" which means as much.
- Sometimes you need to replace the sub part (e.g. search result → OccupiedEntry, iterator → next position). That can be accomplished with a `replace` method taking a closure, but once you have that, you can also simply extract the "Sub" part, replace it with (), and you end up with the same lack of safety as is now

Reply:
I don't understand how you jumped from "keep `NonNull` confined in closure" to

> Having "Sub" as a part of "Self"

so I do not follow any part of that post.

Reply:
Sorry, not clear again. It was about the theme of making it hard to do the wrong thing, and specifically why it doesn't work to embed the "Sub" into the "Self" to make sure the "Sub" is consumed.

So that is not a problem, but the fact that `awaken` is unsafe is a disappointment. I wish the borrow checker would know that no borrows of Sub can be around.

Reply:
If the borrow checker was that smart, we wouldn't need this entire abstraction at all -- we'd just use a normal reborrow for `this`. The entire point of this abstraction, as I see it, is to handle cases where the lifetime of `this` (the thing passed to the closure) is too complicated for the compiler to understand.

Hence my thinking that the closure can be eliminated entirely and `NonNull` be hidden from the public interface.

Reply:
The `NonNull` is hidden now. But how could you eliminate the closure without loosing safety? It allows a last peek at the reference, before it is put to sleep and locked in its room. Otherwise all I see is an unsafe method get_reference_to_the_entire_thing and callers have to promise to call it only once (or more accurately, the same precondition as `awaken`). But then that is exactly `NonNull::as_mut`, right?

Reply:
I don't think so, `as_mut(&mut self)` is much better than that: it borrows the `NonNull` thus guarantees uniqueness of the reference, as long as the `NonNull` is the only copy of the pointer alive. So is `DormantMutRef` simply `NonNull` without unsafe methods? To be continued...

Reply:
I don't think `as_mut(&mut self)` or anything similar is suitable: the returned lifetime is attached to the dormant thing, but we want to store both the new reference and the dormant thing for future use. That only works if we hide the borrowing from the borrow checker, and that means the borrow checker will happily let you invoke `as_mut` as much as you want. One sensible alternative is what we have: hand out a reference only once, in the constructor of the dormant thing.

Reply:
To close the point: indeed the closure can be eliminated, I stuck it in between because I misinterpreted the parent-child relationship of pointers/references.

Reply:
>  So is DormantMutRef simply NonNull without unsafe methods? To be continued...

It's also non-duplicable, which is crucial.
I am not sure what the `as_mut` method you are talking about here is, but maybe you dropped that idea so I don't need to understand it.^^

Reply:
What, you say `NonNull` is Copy and copying is not unsafe? That sounds... unavoidable and used all over the place, like in NodeRef.

Yes, I dropped the idea (or at least made it dormant in the back of my brain), but since you woke it up, it's making a fuss in my brain, complaining about how I spelled "heat" too. I was rambling about `NonNull::as_mut` (which I didn't realize is unsafe itself, but that's obvious now). My first point was the insight that if you create a `NonNull` from a unique reference, and the only method on it you expose is `NonNull::as_mut`, then the borrow checker makes sure everything is safe. But that's not completely true: you need to uphold the borrow of the original reference by accompanying the `NonNull` with phantom data with the right lifetime, so already `DormantMutRef` is not simply a stripped down `NonNull`.

But even if I am right about that, then because of that, `NonNull::as_mut` is unusable as an alternative to returning the reference from `DormantMutRef::new`. For instance, we need to return an `OccupiedEntry` thing all the way back to the user, with the lifetime `'t` of the borrow done by the `map.entry` call, and containing a reference obtained through `NonNull::as_mut`. So we must stretch the lifetime of the reference returned by `NonNull::as_mut` to `'t`, and then we must stretch the lifetime of the argument of `NonNull::as_mut` likewise or the borrow checker would allow overlapping borrows by `NonNull::as_mut`, and then we cannot even pretend to have unique access to the `NonNull` before the end of `'t`, required to "awaken" the original reference. In fact, we cannot invoke anything until the end of `NonNull::as_mut` borrow which is `'t`; in other words, this `NonNull::as_mut` would have the effect of moving. In other words, this `NonNull::as_mut` would _be_ `awaken`.

Or is that unfair of me. `awaken` is unsafe anyway, so just 

Now